x/storage/keeper/keeper.go

```diff
@@ -1464,3 +1479,157 @@ func (k Keeper) getDiscontinueObjectStatus(ctx sdk.Context, objectId types.Uint)
 	store.Delete(types.GetDiscontinueObjectStatusKey(objectId)) //remove it at the same time
 	return types.ObjectStatus(status), nil
 }
+
+func (k Keeper) appendResourceIdForGarbageCollection(ctx sdk.Context, resourceType resource.ResourceType, resourceID sdkmath.Uint) error {
+	if !k.permKeeper.ExistAccountPolicyForResource(ctx, resourceType, resourceID) &&
+		!k.permKeeper.ExistGroupPolicyForResource(ctx, resourceType, resourceID) {
+
+		if resourceType != resource.RESOURCE_TYPE_GROUP ||
+			(resourceType == resource.RESOURCE_TYPE_GROUP && !k.permKeeper.ExistGroupMemberForGroup(ctx, resourceID)) {
+			return nil
+		}
+	}
+	tStore := ctx.TransientStore(k.tStoreKey)
+	var deleteInfo types.DeleteInfo
+	if !tStore.Has(types.CurrentBlockDeleteStalePoliciesKey) {
+		deleteInfo = types.DeleteInfo{
+			BucketIds: &types.Ids{},
+			ObjectIds: &types.Ids{},
+			GroupIds:  &types.Ids{},
+		}
+	} else {
+		bz := tStore.Get(types.CurrentBlockDeleteStalePoliciesKey)
+		k.cdc.MustUnmarshal(bz, &deleteInfo)
+	}
+	switch resourceType {
+	case resource.RESOURCE_TYPE_BUCKET:
+		bucketIds := deleteInfo.BucketIds.Id
+		bucketIds = append(bucketIds, resourceID)
+		deleteInfo.BucketIds = &types.Ids{Id: bucketIds}
+	case resource.RESOURCE_TYPE_OBJECT:
+		objectIds := deleteInfo.ObjectIds.Id
+		objectIds = append(objectIds, resourceID)
+		deleteInfo.ObjectIds = &types.Ids{Id: objectIds}
+	case resource.RESOURCE_TYPE_GROUP:
+		groupIds := deleteInfo.GroupIds.Id
+		groupIds = append(groupIds, resourceID)
+		deleteInfo.GroupIds = &types.Ids{Id: groupIds}
+	default:
+		return types.ErrInvalidResource
+	}
+	tStore.Set(types.CurrentBlockDeleteStalePoliciesKey, k.cdc.MustMarshal(&deleteInfo))
+	return nil
+}
+
+func (k Keeper) PersistDeleteInfo(ctx sdk.Context) {
+	tStore := ctx.TransientStore(k.tStoreKey)
+	if !tStore.Has(types.CurrentBlockDeleteStalePoliciesKey) {
+		return
+	}
+	bz := tStore.Get(types.CurrentBlockDeleteStalePoliciesKey)
+	deleteInfo := &types.DeleteInfo{}
+	k.cdc.MustUnmarshal(bz, deleteInfo)
+
+	// persist current block stale permission info to store if exists
+	if !deleteInfo.IsEmpty() {
+		store := ctx.KVStore(k.storeKey)
+		store.Set(types.GetDeleteStalePoliciesKey(ctx.BlockHeight()), bz)
+		_ = ctx.EventManager().EmitTypedEvents(&types.EventStalePolicyCleanup{
+			BlockNum:   ctx.BlockHeight(),
+			DeleteInfo: deleteInfo,
+		})
+	}
+}
+
+func (k Keeper) GarbageCollectResourcesStalePolicy(ctx sdk.Context) {
+	store := ctx.KVStore(k.storeKey)
+	deleteStalePoliciesPrefixStore := prefix.NewStore(store, types.DeleteStalePoliciesPrefix)
```

Comment:
Too many duplicated code here, I think we can improve it.

Reply:
refactored